### DISCUSSION TITLE
Is it possible to have dynamic injections?

### DISCUSSION THREAD

**Discussion by IndianBoy42** [nvim-treesitter/nvim-treesitter#1833] [Q&A]
I am writing a parser for this language: https://github.com/casey/just, the WIP code is here https://github.com/IndianBoy42/tree-sitter-just

Its essentially a command runner, and it can use any shell/language for the commands to run, it just feeds it line by line to another program (like `bash -c`, `fish -c`). This is controlled by a `set shell := ["bash", "-c"]` command, usually near the top.

Also individual recipes can be preceeded by a shbang like `!#/usr/bin/env python` and the whole contents will be treated like a file with the shbang and executed, so even python/lua other scripting languages can be a part of a Justfile.

How would I go about dealing with this in terms of treesitter parsing and injections? I currently just don't really attempt to parse whats inside the recipes, just consuming words and waiting for a `dedent`. I would like to be able to detect what the `shell` is and what shebangs are used and properly get treesitter to highlight recipes.

What I am thinking is that this can't be done with `injections.scm`, I would have to write some lua code that queries the parse tree and run some logic to find what language a recipe is written in. Then that lua code would have to somehow tell treesitter to highlight that area as `$other_language`, is this last part possible?

**Comment by theHamsta** [comment]
Yes, injection can be dynamic. Each predicate and directive is a Lua function and you can do anything you want in it. Typically, you can even capture the `@language` directly with or do some preprocessing before it. In your case you would most probably work with a per-buffer variable. You would need your custom predicate that sets the `language` metadata to this variable. You could have one query that searches the she bang an just sets the option for the remaining queries or you set the option manually on fileopen.

Someone wrote a Makefile parser sóme time ago that should be a good starting for a just parsers. There're also some open pull request that want to make the injections more configurable. They could also be a good starting point on how the behavior of the injections can be modified.

**Reply by IndianBoy42** [reply]
Where is the documentation for how to create/attach custom predicates and directives? I see now the `@language` capture should be able to be used for the shbangs in individual recipes. But for the global `shell` variable based one I think custom lua code may have to be written.

**Reply by IndianBoy42** [reply]
Okay, @language and @content was enough for both cases. My next question: The recipes are actually not exactly the inner language, there is one extra feature: the string interpolation, no matter how the recipe is run, you can add something like: `{{ variable }}`, and just will substitute in the value of that variable before passing the code to the interpreter/shell. The existence of this seems to mess up highlighting for fish and probably would also do so for some other languages.

How do you think I can deal with that? It seems that inside the interpolation blocks it is highlighted by both the just highlighter and the bash highlighter, is this expected. The other problem is around it, the fish/bash parser can get confused and highlighting is a bit messed up

**Reply by theHamsta** [reply]
There's actually a feature `combined!` of Tree-Sitter that will enable to have certain ranges of the injection region interrupted while the parser state is maintained (they use Ruby templates as an example). The official Tree-Sitter documentation has more on that. It is however not yet fully supported by nvim but we to have it in Future.

**Comment by theHamsta** [comment]
https://github.com/alemuller/tree-sitter-make
